File: Python/Linked List/21_pairs_with_sum_in_doubly_LL.py
"""
Idea:
using 2 pointer approach, one from start and one from end,
 checking if both makes the target value, else updating the pointers
"""

from typing import Optional

from typing import List


class Solution:
    def findPairsWithGivenSum(self, target: int, head: Optional['Node']) -> List[List[int]]:
        # code here
        ans = []
        front = back = head
        while back.next is not None:
            back = back.next

        while front != back and back.next != front:
            if front.data + back.data == target:
                ans.append([front.data, back.data])
                front = front.next
                back = back.prev
            elif front.data + back.data < target:  # if sum < target, incrementing the front pointer
                front = front.next
            else:  # if sum > target, incrementing the front pointer
                back = back.prev
        return ans


# {
# Driver Code Starts
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class doublyLL:
    def __init__(self):
        self.head = None

    def insert(self, tail, data):
        head = self.head

        node = Node(data)

        if not head:
            self.head = node
            return node

        tail.next = node
        node.prev = tail
        return node


def displayList(head):
    while head:
        print(head.data, end=' ')
        pvs = head
        head = head.next
    print()

    while pvs:
        print(pvs.data, end=' ')
        pvs = pvs.prev


if __name__ == '__main__':
    tcs = int(input())

    for _ in range(tcs):
        target = int(input())
        n = int(input())
        arr = [int(x) for x in input().split()]

        dll = doublyLL()

        tail = None

        for nodeData in arr:
            tail = dll.insert(tail, nodeData)

        obj = Solution()
        ans = obj.findPairsWithGivenSum(target, dll.head)
        if (len(ans) == 0):
            print(-1)
        else:

            for i in range(len(ans)):
                print("(" + str(ans[i][0]) + "," + str(ans[i][1]) + ")", end=' ')

            print()

# } Driver Code Ends

File: C++/Linked List/21_pairs_with_sum_in_doubly_LL.cpp
/*
Idea:
using 2 pointer approach, one from start and one from end,
 checking if both makes the target value, else updating the pointers
*/

//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int data;
    Node* next, * prev;
    Node(int val) : data(val), next(NULL), prev(NULL)
    {
    }
};


// } Driver Code Ends
//User function Template for C++


class Solution
{
public:
    vector<pair<int, int>> findPairsWithGivenSum(Node* head, int target)
    {
        vector<pair<int, int>> ans;

        Node* front = head, * back = head;
        while (back->next)
            back = back->next;

        while (front != back && front != back->next)
        {
            if (front->data + back->data == target)
            {
                ans.push_back({ front->data ,back->data });
                front = front->next;
                back = back->prev;
            }
            else if (front->data + back->data < target) // if sum < target, incrementing the front pointer
                front = front->next;
            else  // if sum <> target, decrementing the back pointer
                back = back->prev;
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n, target;
        cin >> target >> n;
        int a;
        cin >> a;
        Node* head = new Node(a);
        Node* tail = head;
        for (int i = 0; i < n - 1; i++)
        {
            cin >> a;
            tail->next = new Node(a);
            tail->next->prev = tail;
            tail = tail->next;
        }
        Solution ob;
        auto ans = ob.findPairsWithGivenSum(head, target);
        if (ans.size() == 0)
            cout << "-1";
        else
        {
            for (int i = 0; i < ans.size(); i++)
            {
                cout << "(" << ans[i].first << "," << ans[i].second << ")"
                    << " ";
            }
        }
        cout << "\n";
    }
    return 0;
}

// } Driver Code Ends